File: R/astar.R
library(igraph)

# Node class to represent each point on the grid
create_node <- function(x, y, g = 0, h = 0, parent = NULL) {
  list(x = x, y = y, g = g, h = h, f = g + h, parent = parent)
}

# Heuristic function based on Manhattan distance
heuristic <- function(node, goal) {
  abs(node$x - goal$x) + abs(node$y - goal$y)
}

# A* function
astar <- function(start, goal, grid) {
  open_set <- list(start)
  closed_set <- list()
  path <- list()
  
  while (length(open_set) > 0) {
    open_set <- open_set[order(sapply(open_set, function(n) n$f))]
    current <- open_set[[1]]
    open_set <- open_set[-1]
    
    if (current$x == goal$x && current$y == goal$y) {
      while (!is.null(current$parent)) {
        path <- c(list(c(current$x, current$y)), path)
        current <- current$parent
      }
      return(do.call(rbind, path))
    }
    
    closed_set <- c(closed_set, list(current))
    
    neighbors <- list(
      create_node(current$x + 1, current$y),
      create_node(current$x - 1, current$y),
      create_node(current$x, current$y + 1),
      create_node(current$x, current$y - 1)
    )
    
    for (neighbor in neighbors) {
      if (neighbor$x < 1 || neighbor$y < 1 || neighbor$x > nrow(grid) || neighbor$y > ncol(grid) || grid[neighbor$x, neighbor$y] == 1) {
        next
      }
      
      if (any(sapply(closed_set, function(n) n$x == neighbor$x && n$y == neighbor$y))) {
        next
      }
      
      neighbor$g <- current$g + 1
      neighbor$h <- heuristic(neighbor, goal)
      neighbor$f <- neighbor$g + neighbor$h
      neighbor$parent <- current
      
      if (!any(sapply(open_set, function(n) n$x == neighbor$x && n$y == neighbor$y && n$f <= neighbor$f))) {
        open_set <- c(open_set, list(neighbor))
      }
    }
  }
  return(NULL)  # No path found
}

# Example usage:
grid <- matrix(c(
  0, 1, 0, 0,
  0, 1, 0, 1,
  0, 0, 0, 1,
  1, 0, 0, 0
), nrow = 4, byrow = TRUE)

start <- create_node(1, 1)
goal <- create_node(4, 4)
path <- astar(start, goal, grid)
print("Path:")
print(path)

File: Ruby/astar.rb
# A* Pathfinding Algorithm in Ruby
# This script finds the shortest path on a grid from a starting point to a goal.

class Node
  attr_accessor :position, :g_cost, :h_cost, :f_cost, :parent

  def initialize(position, g_cost, h_cost, parent = nil)
    @position = position
    @g_cost = g_cost
    @h_cost = h_cost
    @f_cost = g_cost + h_cost
    @parent = parent
  end
end

def heuristic(point, goal)
  # Manhattan distance as heuristic
  (point[0] - goal[0]).abs + (point[1] - goal[1]).abs
end

def astar(grid, start, goal)
  open_set = []
  closed_set = Array.new(grid.size) { Array.new(grid[0].size, false) }
  start_node = Node.new(start, 0, heuristic(start, goal))
  open_set.push(start_node)

  until open_set.empty?
    current_node = open_set.min_by(&:f_cost)
    return reconstruct_path(current_node) if current_node.position == goal

    open_set.delete(current_node)
    closed_set[current_node.position[0]][current_node.position[1]] = true

    neighbors(current_node.position, grid).each do |neighbor_pos|
      next if closed_set[neighbor_pos[0]][neighbor_pos[1]] || grid[neighbor_pos[0]][neighbor_pos[1]] == 1

      tentative_g_cost = current_node.g_cost + 1
      neighbor_node = Node.new(neighbor_pos, tentative_g_cost, heuristic(neighbor_pos, goal), current_node)

      if open_set.none? { |node| node.position == neighbor_pos && node.g_cost <= tentative_g_cost }
        open_set.push(neighbor_node)
      end
    end
  end
  raise 'No path found'
end

def neighbors(position, grid)
  row, col = position
  [[row - 1, col], [row + 1, col], [row, col - 1], [row, col + 1]].select do |r, c|
    r.between?(0, grid.size - 1) && c.between?(0, grid[0].size - 1)
  end
end

def reconstruct_path(node)
  path = []
  while node
    path << node.position
    node = node.parent
  end
  path.reverse
end

# Example usage:
grid = [
  [0, 1, 0, 0],
  [0, 1, 0, 1],
  [0, 0, 0, 1],
  [1, 0, 0, 0]
]

start = [0, 0]
goal = [3, 3]

path = astar(grid, start, goal)
puts "Path: #{path.inspect}"
